tests/callbacks/test_quantization.py

```diff
@@ -139,3 +153,154 @@ def test_quantization_triggers(tmpdir, trigger_fn: Union[None, int, Callable], e
+                # ``torch.quantization.FakeQuantize`` checks observer_enabled[0] == 1.
+                observer_enabled = observer_enabled[0] == 1
+                assert observer_enabled == expected_observer_enabled
+            else:
+                assert torch.equal(observer_enabled, expected_observer_enabled)
+
+    def _check_observer_state(self, should_disable):
+        if should_disable:
+            self._assert_observer_enabled(collections.defaultdict(lambda: False))
+        else:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_fit_start(self):
+        self._fake_quants = tuple(module for module in self.modules() if isinstance(module, FakeQuantizeBase))
+        self._fake_quant_to_initial_state = {
+            fake_quant: copy.deepcopy(fake_quant.state_dict()) for fake_quant in self._fake_quants
+        }
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+    def training_step(self, batch, batch_idx):
+        self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+        result = super().training_step(batch, batch_idx)
+        if self._disable_observers_after_step == self.global_step:
+            self.apply(torch.quantization.disable_observer)
+
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+        return result
+
+    def on_train_start(self):
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+        self._check_observer_state("train" in self._disable_observers)
+
+    def on_train_end(self):
+        if "train" in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_validation_start(self):
+        self._check_observer_state("validate" in self._disable_observers and not self.trainer.sanity_checking)
+
+    def on_validation_end(self):
+        if "validate" in self._disable_observers:
+            if self.trainer.sanity_checking:
+                self._assert_fake_quant_initial_state()
+            else:
+                self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_test_start(self):
+        self._check_observer_state("test" in self._disable_observers)
+
+    def on_test_end(self):
+        if "test" in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_predict_start(self):
+        self._check_observer_state("predict" in self._disable_observers)
+
+    def on_predict_end(self):
+        if "predict" in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+
+class BeforeTrainingCalibration(Callback):
+    def on_train_start(self, trainer, pl_module):
+        pl_module.training_step(next(iter(trainer.train_dataloader)), 0)
```

Comment:
what's this and why is it necessary?

Reply:
Current `MovingAveragePerChannelMinMaxObserver` and `HistogramObserver` need to see at least one batch to infer the shapes of quantization `scale` and `zero_point`. When `train` stage is in `disable_observers`, users are supposed to "initialize" observers before the training loop. In the tests `BeforeTrainingCalibration` will feed one batch to observers before training.